README.md
![](https://socialify.git.ci/Kyomotoi/ATRI/image?description=1&descriptionEditable=A%20project%20for%20ATRI%2C%20Usage%20go-CQHTTP%20%2B%20NoneBot2.&forks=1&issues=1&language=1&logo=https%3A%2F%2Fi.loli.net%2F2020%2F11%2F12%2FYcINCkyp8vK2inD.png&owner=1&pattern=Circuit%20Board&stargazers=1&theme=Light)

### 👋 Here is ATRI - 一个厨力项目

[![time tracker](https://wakatime.com/badge/github/Kyomotoi/ATRI.svg)](https://wakatime.com/badge/github/Kyomotoi/ATRI)
[![](https://img.shields.io/github/license/Kyomotoi/ATRI)](https://www.gnu.org/licenses/gpl-3.0.html)

アトリは、高性能ですから！

为🐧即时聊天工具中复现一只优秀的**功能性**机器人是本项目的目标.

## 📌 声明 | Declarations

**一切开发旨在学习，请勿用于非法用途**

本项目名称、灵感均来自 [ANIPLEX](https://aniplex-exe.com/) 发行的 [ATRI-My Dear Moments-](https://atri-mdm.com/).

本项目中所使用的任何有关 [ATRI](https://atri-mdm.com/) 的图标、LOGO，解释权、著作权均归 [ANIPLEX](https://aniplex-exe.com/). 你可以[在此](https://aniplex-exe.com/guidelines/)查看相关内容.

## ✨ 特性概览 | Features

- 使用 [NoneBot 2](https://v2.nonebot.dev/) 进行项目底层构建.
- 使用 [go-cqhttp](https://go-cqhttp.org/) 作为默认协议端.
- 遵循 [OneBot v11](https://onebot.dev/) 规范.
- 易于上手的 [可视化界面](https://github.com/Kyomotoi/Project-ATRI-Console).
- Coming soon...

## 📱 功能概览 | Services Overview

<details markdown='1'><summary>涩涩</summary>

- 文爱
- 涩图
- 涩图嗅探
- 涩批翻译机
</details>

<details markdown='1'><summary>实用</summary>

- 在线运行代码
- 伪造转发内容
- 以图搜图
- 以图搜番
- ATRI语（加密、解密，改自[`rcnb`](https://github.com/rcnbapp/RCNB.js)）
- 简单骰子
- b站动态订阅
- Twitter 动态订阅
- 谁是卷王 (基于 wakatime)
- 自定义词库（多种方式匹配）
</details>

<details markdown='1'><summary>娱乐</summary>

- 看不懂的笑话
- 今天吃什么
- 老婆！
</details>

<details markdown='1'><summary>其它</summary>

- B站小程序解析
- 状态查看
</details>

<details markdown='1'><summary>TODO</summary>

> 被勾上的即已大致完成，但仍需优化
- [x] 网页控制台
- [ ] Twitter助手
- [x] RSS订阅
- [ ] 冷重启
- [ ] 进裙验证（问题可自定义）
- [ ] 好感度系统
</details>

## 🚀 开始部署 | Getting Started

请参考文档: [部署项目](https://atri.imki.moe/install/installation/)

## 📖 文档 | Documentation
所有公开的信息都可在 [atri.imki.moe](https://atri.imki.moe) 获取. 使用 [MkDocs](https://github.com/mkdocs/mkdocs/) 构建. 

[文档仓库](https://github.com/Kyomotoi/Project-ATRI-Docs)

## ⚡️ 性能表现 | Performance

得益于强大的 [NoneBot 2](https://v2.nonebot.dev/) 和 [go-cqhttp](https://go-cqhttp.org/)，真正意义上实现：全平台运行、运行时资源占用极少.

## 🔊 更新日志 | Changelog

版本更新请参考[此处](changelog.md).

小改动请参考以往的 [commit](https://github.com/Kyomotoi/ATRI/commits/main).

## ❤️ 特别感谢 | Acknowledgments

[Bot Universe](https://github.com/botuniverse): [Onebot标准](https://onebot.dev/)

[Mrs4s](https://github.com/Mrs4s): [go-cqhttp](https://github.com/Mrs4s/go-cqhttp)

[NoneBot](https://github.com/nonebot): [NoneBot2](https://github.com/nonebot/nonebot2)

[Richard Chien](https://github.com/richardchien), [Mnixry](https://github.com/mnixry) and go-cqhttp Developers.

[JetBrains](https://www.jetbrains.com/?from=ATRI): 为本项目提供 [PyCharm](https://www.jetbrains.com/pycharm/?from=ATRI) 等 IDE 的授权<br>
[<img src="https://resources.jetbrains.com/storage/products/company/brand/logos/jb_beam.png" width="150"/>](https://www.jetbrains.com/?from=ATRI)
[<img src="https://resources.jetbrains.com/storage/products/company/brand/logos/PyCharm_icon.png" width="130"/>](https://www.jetbrains.com/pycharm/?from=ATRI)

<details markdown='1'><summary>以及以下朋友们！</summary>

>排名不分先后
- 50861735 11.00 CNY
- 1072324725 17.00 CNY
- AfdianUser_quGy 5.00 CNY
- 1752179928 56.14 CNY
- Mikasa 66.00 CNY
- SkipM4 32.00 CNY
- Chunk7 33.00 CNY
- Wwwwwwalnut 10.00 CNY
- 演变 5.00 CNY
- YuimeC 23.33 CNY
- Ohdmire 20.00 CNY
- TerRALi 23.45 CNY
- 虾仁 21.00 CNY
- Tianli 11.00 CNY
- linx 33.00 CNY
- 屁东 5.00 CNY
- A_Kyuu 200.00 CNY
- Copper_Sulfate 11.00 CNY
- YQ233 11.00 CNY
- Wuid 7.50 CNY
- AfdianUser_veuW 5.00 CNY
- KadiASD 7.50 CNY
- (*°ω°*)ﾉ" 11.45 CNY
- lanx 6.66 CNY
- Azazel_BBF 33.00 CNY
- hongshi 10.00 CNY
- 这是01 45.14 CNY
> Total: 703.17 CNY. 25 sponsors.
</details>

## 👥 贡献 | Contribution

如果你喜欢本项目，可以请我[喝杯红茶](https://afdian.net/@Kyomotoi).

如果你有想法、有能力，欢迎:
- [提交 Issue](https://github.com/Kyomotoi/ATRI/issues)
- [提交 Pull request](https://github.com/Kyomotoi/ATRI/pulls)
- [在交流群内进行反馈](https://jq.qq.com/?_wv=1027&k=WoAAYXbJ)

在提交 Issue、Pull request 前，请先阅读[此内容](https://atri.imki.moe/developer/overview/).

## 📄 许可 | License

本项目使用 [GPLv3](https://www.gnu.org/licenses/gpl-3.0.html).

意味着你可以运行本项目，并向你的用户提供服务，如后续有对本项目源码的修改，你需要向用户公开修改后的此项目的源码.


changelog.md
> 此处仅为记录重大更新，修复 BUG/以及其它 请关注[`GitHub commits`](https://github.com/Kyomotoi/ATRI/commits/main)

## Coming soon...
- Next: `YHN-001-A06`

---

## Feb 7, 2022

> 接下来该把精力放到 [`RouterForGuild`](https://github.com/Kyomotoi/RouterForGuild) 了

- 更新版本至: `YHN-001-A05.fix1`
- 新增:
    - 插件：[`帮助`](user/plugin-help.md)
        - 新增本文档地址
    - 插件：[`涩图`](user/plugin-setu.md)
        - 添加了 `冲后感`
    - `config.yml`:
        - `InlineGoCQHTTP`: `download_version` 用于指定内置 `gocqhttp` 下载、运行版本
- 其它:
    - 更新:
        - 所有需要冷却的插件均换上了 `NoneBot v2.0.0 beta1` 内置的 `Cooldown`<br>（原来的由于未知原因已失效）
        - 优化本项目 `README`
    - 修复:
        - 插件：[`老婆`](user/plugin-wife.md)
            - QQ号指向错误
            - 由于**原子性**，后续bug可能会频发，正考虑是否删除
        - 部分插件索取图片相关部分未完全适配 `NoneBot v2.0.0 beta1`
        - 内置 `gocqhttp` 指向版本 `latest` 下载**可能**提示 `404 Not Found`

---

## Jan 31, 2022

> 新年快乐！

- 更新版本至: `YHN-001-A05`
- 新增:
    - 全面适配 `NoneBot v2.0.0 beta1`
    - 插件：[`broadcast`](user/plugin-broadcast.md)
        - 现在可以向机器人所在群推送同一消息
    - 插件：[`essential`](user/plugin-essential.md)
        - 现在可以接受、拒绝撤回消息推送
    - 内置 [`gocqhttp`](https://github.com/mnixry/nonebot-plugin-gocqhttp)
        - 现在可以连同机器人一同启动了
        - 同时包含了可视化gocq控制台
            - 可快捷、快速添加新账号
            - 生产环境资源可视化
    - [单元测试](/test)，加快了机器人开发上线前自检的速度
    - `config.yml` 新增以下内容:
        ```yaml
        InlineGoCQHTTP:
        accounts: # 可多个账号，具体请参考文档
            - uin: 1234567890
            password: ""
            protocol: 3
        
        download_version: "latest"

        Setu:
        reverse_proxy: true # 请参考文档
        reverse_proxy_domain: "i.pixiv.re"
        ```
        - `config` 中 `proxy` 真正意义上生效
- 其它:
    - 移除插件: `cause`
        - 原因：过于无趣，遭大多用户反对，且源API已失效，不想再替换新的第三方API

---

## Oct 24, 2021

- 更新版本至: `YHN-001-A04`
- 新增：
    - nsfw检测（主动/被动）又名 `涩图嗅探`
    - 可选代理
- 修复：
    - plugin/chat 在 nb2-a14+ 版本 finish 内为空时会报错
- 其他：
    - 对定时任务进行中文命名
- 移除:
    - 前端界面（赶工而导致成品炸裂）

---

## Jul 31, 2021

- 新增:
    - 前端（单主页）
    - 和维护者贴贴w

---

## Jul 8, 2021

-  更新版本至: `YHN-001-A03`，请关注commit: [传送](https://github.com/Kyomotoi/ATRI/commit/be2747e4d4b820ca0f1f988d3b77a628da26fe7b)

---

## May 4, 2021

- 新增:
    - 涩图
    - 群老婆！

---

## Apr 11, 2021

- 正式重构完成

---

## 更早记录

> 本项目始于 2020 年 5月，原项目名: `Aya`，后更改为现在的名称

- 请自行查看[`GitHub commits`](https://github.com/Kyomotoi/ATRI/commits/main)


html.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
'''
@File    :   app.py
@Time    :   2020/10/24 00:25:34
@Author  :   Kyomotoi
@Contact :   kyomotoiowo@gmail.com
@Github  :   https://github.com/Kyomotoi
@License :   Copyright © 2018-2020 Kyomotoi, All Rights Reserved.
'''
__author__ = 'kyomotoi'

import json
import time
import sqlite3
from enum import Enum
from pathlib import Path
from fastapi import FastAPI
from starlette.requests import Request
from starlette.templating import Jinja2Templates

from utils.utils_request import aio_get_bytes


# 急着去上学写乱了下次一定改别骂了别骂了我爪巴爪巴爪巴
# orz orz orz orz orz


app = FastAPI()
tmp_API = Jinja2Templates(directory='ATRI/data/data_HTML/api')
tmp_HELP = Jinja2Templates(directory='ATRI/data/data_HTML/user')


async def get_ip_info(ip: str) -> None:
    URL = await aio_get_bytes(f"http://ip-api.com/json/{ip}?lang=zh-CN&fields=status,message,continent,continentCode,country,countryCode,region,regionName,city,district,zip,lat,lon,timezone,isp,org,as,asname,reverse,mobile,proxy,hosting,query")
    info = json.loads(URL)
    return info

def check_visitors(query: str):
    try:
        with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'banip' / 'banip.json', 'r') as f:
            data = json.load(f)
    except:
        data = {}
    
    if query in data:
        return False
    else:
        return True

async def load_setu(name: str):
    file_type = 'normal'
    s_type = 'normal'

    if name == "normal":
        file_type = file_setu_normal
        s_type = "normal"

    elif name == "nearR18":
        file_type = file_setu_nearR18
        s_type = "nearR18"

    elif name == "R18":
        file_type = file_setu_R18
        s_type = "r18"

    else:
        return {
                "code": 404,
                "data": "",
                "msg": "Can't find aims type!"
                }

    con = sqlite3.connect(file_type)
    cur = con.cursor()
    info = cur.execute(f'SELECT * FROM {s_type} ORDER BY RANDOM() limit 1;')
    for i in info:
            pid = i[0]
            title = i[1]
            tags = i[2]
            account = i[3]
            name = i[4]
            u_id = i[5]
            user_link = i[6]
            img = i[7]
            print('success!')
            con.commit()
            con.close()
            return {
                    "code": 200,
                    "pid": pid,
                    "title": title,
                    "tags": tags,
                    "account": account,
                    "name": name,
                    "u_id": u_id,
                    "user_link": user_link,
                    "img": img
                    }


file_setu_normal = Path('.') / 'ATRI' / 'data' / 'data_Sqlite' / 'setu' / 'normal.db'
file_setu_nearR18 = Path('.') / 'ATRI' / 'data' / 'data_Sqlite' / 'setu' / 'nearR18.db'
file_setu_R18 = Path('.') / 'ATRI' / 'data' / 'data_Sqlite' / 'setu' / 'r18.db'

con = sqlite3.connect(file_setu_normal) # setu-normal
cur = con.cursor()
cur.execute("select * from normal")
data_normal = len(cur.fetchall())
con.close()

con = sqlite3.connect(file_setu_nearR18) # setu-nearR18
cur = con.cursor()
cur.execute("select * from nearR18")
data_nearR18 = len(cur.fetchall())
con.close()

con = sqlite3.connect(file_setu_R18) # setu-r18
cur = con.cursor()
cur.execute("select * from r18")
data_r18 = len(cur.fetchall())
con.close()

class SetuModelName(str, Enum):
    noraml = "normal"
    nearR18 = "nearR18"
    R18 = "r18"


localtime = time.asctime(time.localtime(time.time()))


@app.get("/api")
async def index(request: Request):
    ip = request.client.host
    port = request.client.port
    query = str(ip) + ":" + str(port)
    info = await get_ip_info(ip)

    try:
        with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_all.json', 'r') as f:
            data_all = json.load(f)
    except:
        data_all = {}
    
    try:
        with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_index.json', 'r') as f:
            data_index = json.load(f)
    except:
        data_index = {}
    
    info = await get_ip_info(ip)
    data_all[f"{time.time()}"] = [query, info, "index"]
    data_index[f"{time.time()}"] = [query, info]

    with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_all.json', 'w') as f:
        f.write(json.dumps(data_all))
        f.close()
    
    with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_index.json', 'w') as f:
        f.write(json.dumps(data_index))
        f.close()

    return tmp_API.TemplateResponse('index.html',
                                    {'request': request,
                                    'data_normal': data_normal,
                                    'data_nearR18': data_nearR18,
                                    'data_r18': data_r18,
                                    'times_all': "完善中",
                                    'date_now': localtime,
                                    'times_now': "完善中",
                                    'date_yesterday': "完善中",
                                    'times_yesterday': "完善中",
                                    'date_before_yesterday': "完善中",
                                    'times_before_yesterday': "完善中",
                                    'info_ip': query,
                                    'info_continent': info["continent"],
                                    'info_country': info["country"],
                                    'info_regionName': info["regionName"],
                                    'info_city': info["city"],
                                    'info_lat': info["lat"],
                                    'info_lon': info["lon"],
                                    'info_timezone': info["timezone"],
                                    'info_isp': info["isp"],
                                    'info_as': info["as"],
                                    'info_asname': info["asname"]
                                    }
                                    )

@app.get("/api/setu/{s_type}")
async def get_setu(s_type: str, request: Request):
    ip = request.client.host
    port = request.client.port
    query = str(ip) + ":" + str(port)

    try:
        with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_all.json', 'r') as f:
            data_all = json.load(f)
    except:
        data_all = {}
    
    try:
        with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_setu.json', 'r') as f:
            data_setu = json.load(f)
    except:
        data_setu = {}
    
    info = await get_ip_info(ip)
    data_all[f"{time.time()}"] = [query, info, "setu"]
    data_setu[f"{time.time()}"] = [query, info]

    with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_all.json', 'w') as f:
        f.write(json.dumps(data_all))
        f.close()
    
    with open(Path('.') / 'ATRI' / 'data' / 'data_HTML' / 'api' / 'data' / 'times' / 'api_setu.json', 'w') as f:
        f.write(json.dumps(data_setu))
        f.close()

    if check_visitors(query):
        if s_type == SetuModelName.noraml:
            return await load_setu(s_type)
        
        elif s_type == SetuModelName.nearR18:
            return await load_setu(s_type)
        
        elif s_type == SetuModelName.R18:
            return await load_setu(s_type)

    else:
        return {
                "code": 403,
                "data": "",
                "msg": "请停止非法行为！如再继续，达到次数将永久Ban IP！"
                }

if __name__ == '__main__':
    file_info = Path('.') / 'key.json'
    with open(file_info, 'r') as f:
        file_info = json.load(f)
    import uvicorn
    uvicorn.run(app, host=file_info["html"]["ip"], port=file_info["html"]["port"])